finality-aleph/src/finalization.rs

```diff
@@ -77,194 +80,133 @@ where
-
-        for _ in 1..=n {
-            let block = client
-                .new_block(Default::default())
-                .unwrap()
-                .build()
-                .unwrap()
-                .block;
-
-            blocks.push(block.header.hash());
-            futures::executor::block_on(client.import(BlockOrigin::Own, block)).unwrap();
-        }
-
-        blocks
-    }
-
-    #[test]
-    fn should_finalize_for_descendant() {
-        let mut client = Arc::new(TestClientBuilder::new().build());
-
-        let n = 5;
-        let blocks = create_chain(&mut client, n as u64);
-        for i in 0..n {
-            for j in i..n {
-                let maybe_data = should_finalize(
-                    blocks[i],
-                    AlephData::new(blocks[j], j as u64),
-                    client.as_ref(),
-                    100u64,
-                );
-                let correct_result = if i == j {
-                    None
-                } else {
-                    Some(AlephData::new(blocks[j], j as u64))
-                };
-                assert!(maybe_data == correct_result);
+/// Given the `last_finalized` block and `AlephData` `new_data` this returns `Some(block)`
+/// if `new_data` should trigger finalizing `block` or `None` if no new block should be finalized.
+pub(crate) fn block_to_finalize_from_data<B, C>(
+    last_finalized: &BlockHashNum<B>,
+    new_data: AlephData<B>,
+    client: &C,
+    session_boundaries: SessionBoundaries<B>,
+) -> Option<BlockHashNum<B>>
+where
+    B: Block,
+    C: HeaderBackend<B>,
+{
+    match new_data {
+        AlephData::Empty => None,
+        AlephData::HeadProposal(proposal) => {
+            if !is_proposal_available_and_correct(&proposal, client, session_boundaries) {
+                return None;
             }
-        }
-    }
-
-    #[test]
-    fn should_finalize_for_non_descendant() {
-        let mut client = Arc::new(TestClientBuilder::new().build());
-
-        let n = 5;
-        let blocks = create_chain(&mut client, n as u64);
-
-        for i in 0..=n {
-            for j in 0..i {
-                let maybe_data = should_finalize(
-                    blocks[i],
-                    AlephData::new(blocks[j], j as u64),
-                    client.as_ref(),
-                    100u64,
-                );
-                assert!(maybe_data.is_none());
+            // At this point we know that data validation went through, which means that the blocks are available
+            // and the whole `new_data` is honest and correct. Note that there might be cases in which some
+            // other node actually failed data validation for `new_data`. This could have happened in two cases:
+            // 1) Data availability let new_data through even though the `proposal` `new_data.branch.last()` was not available
+            // because it could infer that the `proposal` couldn't be finalized anyway (some conflicting block was finalized).
```

Comment:
sorry, I do not get it - we do perform check of availability of the whole branch

Reply:
Well, this was super confusing and simply incorrect. So it's my bad :/
I fixed the comment, now it should make sense...